compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,939 @@
+    //     !4 = !{i64 0, !"_ZTSFvu27NvNtC1234_5crate6Trait13fooIu22NtC1234_5crate7Struct1Iu3i32ES_EE"}
+    //
+    // The reason for not using v0's extended form of paths is to use a consistent and simpler
+    // encoding, as the reasoning for using it isn't relevand for type metadata identifiers (i.e.,
+    // keep symbol names close to how methods are represented in error messages). See
+    // https://rust-lang.github.io/rfcs/2603-rust-symbol-name-mangling-v0.html#methods.
+    let mut s = String::new();
+
+    // Start and namespace tags
+    let mut def_path = tcx.def_path(def_id);
+    def_path.data.reverse();
+    for disambiguated_data in &def_path.data {
+        s.push('N');
+        s.push_str(match disambiguated_data.data {
+            hir::definitions::DefPathData::Impl => "I", // Not specified in v0's <namespace>
+            hir::definitions::DefPathData::ForeignMod => "F", // Not specified in v0's <namespace>
+            hir::definitions::DefPathData::TypeNs(..) => "t",
+            hir::definitions::DefPathData::ValueNs(..) => "v",
+            hir::definitions::DefPathData::ClosureExpr => "C",
+            hir::definitions::DefPathData::Ctor => "c",
+            hir::definitions::DefPathData::AnonConst => "k",
+            hir::definitions::DefPathData::ImplTrait => "i",
+            hir::definitions::DefPathData::CrateRoot
+            | hir::definitions::DefPathData::Use
+            | hir::definitions::DefPathData::GlobalAsm
+            | hir::definitions::DefPathData::MacroNs(..)
+            | hir::definitions::DefPathData::LifetimeNs(..) => {
+                bug!("encode_ty_name: unexpected `{:?}`", disambiguated_data.data);
+            }
+        });
+    }
+
+    // Crate disambiguator and name
+    s.push('C');
+    s.push_str(&to_disambiguator(tcx.stable_crate_id(def_path.krate).to_u64()));
+    let crate_name = format!("{}", tcx.crate_name(def_path.krate));
+    s.push_str(&format!("{}", crate_name.len()));
+    s.push_str(&crate_name);
+
+    // Disambiguators and names
+    def_path.data.reverse();
+    for disambiguated_data in &def_path.data {
+        let num = disambiguated_data.disambiguator as u64;
+        if num > 0 {
+            s.push_str(&to_disambiguator(num));
+        }
+
+        let name = format!("{}", disambiguated_data.data);
+        s.push_str(&format!("{}", name.len()));
+
+        // Prepend a '_' if name starts with a digit or '_'
+        if name.chars().nth(0).unwrap().is_digit(10) || name.chars().nth(0).unwrap() == '_' {
+            s.push('_');
+        }
+
+        s.push_str(&name);
+    }
+
+    s.to_string()
+}
+
+/// Encodes a ty:Ty using the Itanium C++ ABI with vendor extended type qualifiers and types for
+/// Rust types that are not used at the FFI boundary.
+fn encode_ty<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    ty: Ty<'tcx>,
+    dict: &mut FxHashMap<DictKey<'tcx>, usize>,
+    options: EncodeTyOptions,
+) -> String {
+    let mut typeid = String::new();
```

Comment:
write to outer buffer, and change usage places
```suggestion
fn encode_ty<'tcx>(
    tcx: TyCtxt<'tcx>,
    ty: Ty<'tcx>,
    dict: &mut FxHashMap<DictKey<'tcx>, usize>,
    options: EncodeTyOptions,
    typeid: &mut String,
) {
```

Reply:
Uhh, a lot of similar things:
use `write!(s,"{}", ...)` instead of `s.push_str(&format!(...))` so clippy(https://rust-lang.github.io/rust-clippy/master/#format_push_string) will be happy
and most of string returning non public fns can write in provided buffer, to reduce allocs.

Reply:
While I agree with the push_str+format optimization, we must be careful with micro-optimizations sacrificing API consistency and ergonomics, code readability and maintainability, but not bringing any significant benefits. I'll keep an eye out for and work on them if any issues arise.